hat/app/org/hatdex/hat/api/controllers/Authentication.scala

```diff
@@ -223,4 +225,57 @@ class Authentication @Inject() (
         Future.successful(Unauthorized(Json.toJson(ErrorMessage("Invalid Token", "Token does not exist"))))
     }
   }
+
+  /*
+    BEGIN: hat-claim
+    Entire Section below is on HAT Claim.
+    TODO: Refactor to New File?
+   */
+  /**
+   * Sends an email to the owner with a link to claim the hat
+   */
+  def claim(): Action[ApiClaimHatRequest] = UserAwareAction.async(parsers.json[ApiClaimHatRequest]) { implicit request =>
+    val claimHatRequest = request.body
+    val email = request.dynamicEnvironment.ownerEmail
+    val response = Ok(Json.toJson(SuccessResponse("You will shortly receive an email with claim instructions")))
+
+    if (claimHatRequest.email == request.dynamicEnvironment.ownerEmail) {
+      usersService.listUsers.map(_.find(_.roles.contains(Owner()))).flatMap {
+        case Some(user) =>
+          applicationsService.applicationStatus()(request.dynamicEnvironment, user, request).flatMap { applications =>
+            val maybeApplication = applications.find(_.application.id.equals(claimHatRequest.applicationId))
+            val maybeEmailDetails = maybeApplication.map(app => {
+              (app.application.info.name, app.application.info.graphics.logo.normal)
+            })
+
+            // check existing Token  where isSign = true
```

Comment:
Logic for handling 3 possible states of the claim token is still pending.

Reply:
The 3 possible states of the Token are
1. Unclaimed and date valid
2. Unclaimed and Expired
3. Claimed.

***
State 3 is impossible. The HAT doesn't know it has been claimed.
This is because the token is deleted from db when it is `consumed`, so if a search of email and isSignup = true returns empty, it means either
a. an unclaimed hat has been claimed.
b. it's a normal hat.

Of course, we can add more states into the db, but....

Reply:
You are right. The most semantically correct solution would be to replace deletion procedure with consumedAt column and do null checks to determine the status of the token. However, the complexity of such refactor is not fully known to me as it would also impact the whole password reset flow.
What we can try to do for the signup token, instead of deleting it at consumption, simply set the expiry time to the timestamp of consumption. In such a case, the scenarios 2 and 3 become indistinguishable, but it shouldn't be a problem - the state caused by scenarios 2 wouldn't last long because the HAT would get eventually deleted. All the other cases use cases should be still implementable.

Reply:
Done. - "instead of deleting it at consumption, simply set the expiry time to the timestamp of consumption"